lib/widgets/action_sheet.dart

```diff
@@ -377,6 +387,106 @@ class UserTopicUpdateButton extends ActionSheetMenuItemButton {
   }
 }
 
+class ResolveUnresolveButton extends ActionSheetMenuItemButton {
+  ResolveUnresolveButton({
+    super.key,
+    required this.topic,
+    required this.someMessageIdInTopic,
+    required super.pageContext,
+  }) : _actionIsResolve = !topic.isResolved;
+
+  final bool _actionIsResolve;
+
+  /// The topic that the action sheet was opened for.
+  ///
+  /// There might not currently be any messages with this topic;
+  /// see dartdoc of [ActionSheetMenuItemButton].
+  final TopicName topic;
+
+  /// The message ID that was passed when opening the action sheet.
+  ///
+  /// The message with this ID might currently not exist,
+  /// or might exist with a different topic;
+  /// see dartdoc of [ActionSheetMenuItemButton].
+  final int someMessageIdInTopic;
+
+  @override
+  IconData get icon => _actionIsResolve ? ZulipIcons.check : ZulipIcons.check_remove;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return _actionIsResolve
+      ? zulipLocalizations.actionSheetOptionResolveTopic
+      : zulipLocalizations.actionSheetOptionUnresolveTopic;
+  }
+
+  @override void onPressed() async {
+    final zulipLocalizations = ZulipLocalizations.of(pageContext);
+    final store = PerAccountStoreWidget.of(pageContext);
+    final message = store.messages[someMessageIdInTopic] as StreamMessage?;
+
+    if (message != null && !message.topic.isSameAs(topic)) {
+      // TODO also interrupt on a channel move, with a specific message?
+
+      // The message's topic doesn't match what it was when we opened the
+      // action sheet. So either the resolve/unresolve action was already done
+      // or the topic was renamed; either way, tell the user and don't make a
+      // request.
```

Comment:
This is a good thought. I haven't gone through the PR yet, but we don't already do this for other topic actions like changing topic visibility. Maybe that can be a follow-up issue? Some bits from this if-statement might be reusable in that case.

Reply:
Added a comment about this.